### PR TITLE
[ROCm][Deepseek] dsv3.2 further optimization

### PR DESCRIPTION
**Pull request by ganyi1996ppo**
## Purpose
This PR will replace https://github.com/vllm-project/vllm/pull/32649. 
The main optimization include the following part:
- shuffle indexer cache layout for kv cache update and fetch 
- rocm gluon version of ·paged_mqa_logits` integration
- build ragged metadata for mla at metadata_builder instead of runtime
- fp8 sparse mla support 

please help to review cc @tjtanaa 
## Test Plan
gsm8k score:
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|    20|exact_match|↑  |0.9477|±  |0.0061|
|     |       |strict-match    |    20|exact_match|↑  |0.9462|±  |0.0062|
```
## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py [hunk 1]
```diff
@@ -14,7 +14,7 @@
 from vllm.v1.attention.ops.common import pack_seq_triton, unpack_seq_triton
 
 if current_platform.is_cuda_alike():
```
Comment (tjtanaa) [inline vllm/v1/attention/ops/rocm_aiter_mla_sparse.py]: in this case, we should remove this whole if else statement 

```diff
- if current_platform.is_cuda_alike():
-    pass
```

File: vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py [hunk 2]
```diff
@@ -366,23 +553,30 @@ def forward_mqa(
...
+        if fp8_attention:
+            original_q_shape = q.shape
+            kv_c_and_k_pe_cache = kv_c_and_k_pe_cache.view(current_platform.fp8_dtype())
+            q, _ = ops.scaled_fp8_quant(q.view(q.shape[0], -1), layer._q_scale)
+            q = q.view(original_q_shape)
         mla_padded_q = AiterMLAHelper.get_mla_padded_q(self.num_heads, q)
-        attn_out = self._forward_bf16_kv(
-            mla_padded_q, kv_c_and_k_pe_cache, topk_indices_global, attn_metadata
-        )
+        attn_out = self._forward_mla(layer, mla_padded_q, kv_c_and_k_pe_cache, attn_metadata)
```
Comment (tjtanaa) [inline vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py]: @ganyi1996ppo please fix the pre-commit

File: vllm/v1/attention/backends/mla/rocm_aiter_mla.py [hunk 3]
```diff
@@ -396,6 +396,7 @@ class AiterMLAHelper:
     """
 
     _AITER_MIN_MLA_HEADS: Final = 16
+    _AITER_UNSUPPORTED_HEADS = [32]
```
Comment (tjtanaa) [inline vllm/v1/attention/backends/mla/rocm_aiter_mla.py:399]: Is it just head size 32 that has issue?

So can I understand as

head size 16 is fine; head size 32 is NOT supported;
head size 64 is fine; head size 128 is fine as well, etc .

File: vllm/v1/attention/backends/mla/indexer.py [hunk 4]
```diff
@@ -122,7 +122,7 @@ def get_name() -> str:
 
     @staticmethod
     def get_supported_kernel_block_sizes() -> list[int | MultipleOf]:
-        return [1 if current_platform.is_rocm() else 64]
+        return [64]
```
Comment (tjtanaa) [inline vllm/v1/attention/backends/mla/indexer.py]: Must the indexser page size also match the kvcache kernel block size?

Currently the kernel block size is `[1, 64]`, yet the indexer is only supporting `64`

File: vllm/v1/attention/backends/mla/rocm_aiter_mla.py (hunk 3 above)
Comment (ganyi1996ppo) [reply vllm/v1/attention/backends/mla/rocm_aiter_mla.py:399]: Looks like so, I found when head_size equals to 32, the aiter will automatically pick the, as the symbol name suggested, 16head implementation and cause the illegal memory access. I'm not sure if other shape have the same issue, I only tested tp8 and tp4 case

File: vllm/model_executor/models/deepseek_v2.py [hunk 5]
```diff
@@ -671,30 +671,16 @@ def forward(
...
-        # so we need to reshape back to token-flattened shapes
-        q_pe = q_pe.reshape(-1, self.n_head, self.rope_dim)
-        k_pe = k_pe.reshape(-1, 1, self.rope_dim)
-
-        # `rotary_emb` is shape-preserving; `q_pe` is already
-        # [num_tokens, n_head, rope_dim].
-        q = torch.cat([q_pe, q_nope], dim=-1)
-        # `k_pe` is [num_tokens, 1, rope_dim] (MQA).
-        k = torch.cat([k_pe.squeeze(-2), k_nope], dim=-1)
+        rotary_emb(
```
Comment (tjtanaa) [inline vllm/model_executor/models/deepseek_v2.py]: I am not familiar with the changes in this part of the code. Do you have a way to validate on CUDA?

Comment (tjtanaa) [reply vllm/model_executor/models/deepseek_v2.py]: Else, if possible, create a ROCm specific path instead

File: vllm/v1/attention/backends/mla/indexer.py (hunk 4 above)
Comment (ganyi1996ppo) [reply vllm/v1/attention/backends/mla/indexer.py]: Actually, the mla kernel doesn't care about block size in rocm sparse mla, since the slot is picked up by indexer. I suggest we keep the block-size the same as nvidia, even we might also support block-size 1

File: vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py (hunk 2 above)
Comment (ganyi1996ppo) [reply vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py]: Sure, fixed

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py (hunk 1 above)
Comment (ganyi1996ppo) [reply vllm/v1/attention/ops/rocm_aiter_mla_sparse.py]: Removed

File: vllm/model_executor/models/deepseek_v2.py (hunk 5 above)
Comment (ganyi1996ppo) [reply vllm/model_executor/models/deepseek_v2.py]: another condition branch is added

**Review by tjtanaa — APPROVED** [review]
